Sources/PenguinStructures/AnyValue.swift

```diff
@@ -0,0 +1,218 @@
+  ///
+  /// - Postcondition: where `a` is the created instance, `a.storedType == T.self`, and `a[T.self]`
+  ///   is equivalent to `x`.
+  public init<T>(_ x: T) {
+    if MemoryLayout<T>.fitsExistentialInlineBuffer { storage = x}
+    else { storage = Box(x) as BoxBase }
+  }
+
+  /// The layout of an `Any`.
+  private typealias AnyLayout = (Int, Int, Int, storedType: Any.Type)
+
+  /// The type of the value stored directly in `storage`.
+  private var directlyStoredType: Any.Type {    
+    return Swift.withUnsafePointer(to: self) {
+      UnsafeRawPointer($0).assumingMemoryBound(to: AnyLayout.self)[0].storedType
+    }
+  }
+
+  /// The type of the stored value.
+  public var storedType: Any.Type {
+    let d = directlyStoredType
+    if d != BoxBase.self { return d }
+    return self[unsafelyAssuming: Type<BoxBase>()].storedType
+  }
+
+  /// Returns a pointer to the `T` which is assumed to be stored in `self`.
+  private func pointer<T>(toStored _: Type<T>) -> UnsafePointer<T> {
+    .init(Swift.withUnsafePointer(to: self) { Handle<T>($0) }.pointer)
+  }
+
+  // Existentials that store large types have their own built-in box, but we are unable to detect
+  // uniqueness of an existential box, and they seem to be reallocated unpredictably in some cases.
+  // Instead, we box manually and only use the `Any` for its inline storage.
+  
+  /// A base class for boxing types that don't fit the inline buffer.
+  private class BoxBase {
+    /// The type stored in this box.
+    final let storedType: Any.Type
+
+    /// Creates an instance with the given `storedType`
+    fileprivate init(storedType: Any.Type) {
+      self.storedType = storedType
+    }
+
+    /// Returns the boxed value.
+    fileprivate var asAny: Any { fatalError("override me") }
+  }
+  
+  private final class Box<T>: BoxBase {
+    /// The boxed value
+    var value: T
+
+    /// Creates an instance storing `value`
+    init(_ value: T) {
+      assert(!(value is BoxBase), "unexpected double-boxing!")
+      self.value = value
+      super.init(storedType: T.self)
+    }
+
+    /// Returns the boxed value.
+    fileprivate override var asAny: Any { value }
+  }
+  
+  /// Returns a pointer to the `T` which is assumed to be stored in `self`.
+  private mutating func mutablePointer<T>(toStored _: Type<T>) -> UnsafeMutablePointer<T> {
+    if !MemoryLayout<T>.fitsExistentialInlineBuffer {
+      if !isKnownUniquelyReferenced(&self[unsafelyAssuming: Type<BoxBase>()]) {
+        storage = Box(self[unsafelyAssuming: Type<T>()]) as BoxBase
+      }
+    }
+    return withUnsafeMutablePointer(to: &self) { Handle<T>($0) }.pointer
+  }
+  
+  /// A tool for accessing the `T` stored in the `Any`.
+  private struct Handle<T> {
+    // Warning: this type may look like needless complication but efficient optimization for
+    // AnyValue is extremely sensitive to the shape of its implementation and I've been unable to
+    // eliminate it without hurting the generated code.
```

Comment:
This is valuable, but I think insufficient for reliable maintenance over time. Please include information on what benchmarks and/or SIL/assembly tests a maintainer should look at in order to ensure modifications (or new toolchains) don't cause regressions.

Reply:
Made it go away without changing any of the generated disassembly I was inspecting.